### PR TITLE
AKCORE-107: Complete handleShareAcknowledgeRequest in KafkaApis.scala

### PR DESCRIPTION
**Pull request by chirag-wadhwa5**
This PR contains the updated handleShareAcknolwedge API integrated with the share sessions.

Reference: [AKCORE-107](https://confluentinc.atlassian.net/jira/software/c/projects/AKCORE/boards/2050/backlog?selectedIssue=AKCORE-107)

[AKCORE-107]: https://confluentinc.atlassian.net/browse/AKCORE-107?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 1]
```diff
@@ -467,6 +467,36 @@ public ShareFetchContext newContext(String groupId, Map<TopicIdPartition,
         return context;
     }
 
+    public void acknowledgeSessionUpdate(String groupId, ShareFetchMetadata reqMetadata) {
+        if (reqMetadata.epoch() == ShareFetchMetadata.INITIAL_EPOCH) {
+            // ShareAcknowledge Request cannot have epoch as INITIAL_EPOCH (0)
+            throw Errors.INVALID_SHARE_SESSION_EPOCH.exception();
+        } else if (reqMetadata.epoch() == ShareFetchMetadata.FINAL_EPOCH) {
+            ShareSessionKey key = shareSessionKey(groupId, reqMetadata.memberId());
+            if (cache.remove(key) != null) {
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/SharePartitionManager.java:476]: Doesn't removing from the cache require synchronized too?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -1486,13 +1438,15 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     val shareFetchData = shareFetchRequest.shareFetchData(topicNames)
     val forgottenTopics = shareFetchRequest.forgottenTopics(topicNames)
+    var cachedTopicPartitions : util.List[TopicIdPartition] = null
 
-    var cachedTopicPartitions : util.List[TopicIdPartition] = new util.ArrayList[TopicIdPartition]
-    try {
-      cachedTopicPartitions = sharePartitionManager.cachedTopicIdPartitionsInShareSession(groupId, Uuid.fromString(memberId))
-    } catch {
+    if (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH) {
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1443]: I don't quite see what's special about FINAL_EPOCH here.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 3]
```diff
@@ -4563,111 +4625,95 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+
+    val authorizedTopics = authHelper.filterByAuthorized(
+      request.context,
+      READ,
+      TOPIC,
+      topicIdPartitionSeq
+    )(_.topicPartition.topic)
+
+    val acknowledgeResult : mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]
+    = handleAcknowledgements(request, topicNames, sharePartitionManager, authorizedTopics, groupId, memberId, false)
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:4700]: Indentation of line 4700 is off.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. A few minor comments, but overall looks about right.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (chirag-wadhwa5) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1443]: Thanks for reviewing the PR. We actually don't need the variable `cachedTopicPartitions` at all until it's a FINAL request. That's why the check was added here. When it's the final epoch and the consumer does not acknowledge all the acquired records, then the unacknowledged but acquired records need to be released. cachedTopicPartitions gives a list of all the topic partitions from which the consumer consumed. Later in the end of the function, this list is used to release all acquired records from the same memberId in all these topic partitions.

**Comment by chirag-wadhwa5** [comment]
I have made the required changes in the latest commit, except one. I have left a reply to the comment

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (AndrewJSchofield) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1449]: I think this is worth a comment. You can do that on another commit in this PR, or a future PR.

**Review by AndrewJSchofield — APPROVED** [review]
Thanks for the PR. Approved.

**Pull request merged**